mantidproject/mantid
# Implement ChangeTimeZero algorithm to shift times in logs and neutron timestamps

Implements ticket [#11733](http://trac.mantidproject.org/mantid/ticket/11733)

**Context:**

In order to implement time slicing for event data in SANS we need to alter the time stamps for both the logs and the neutron data. This algorithm is part of this solution. It takes an input workspace, outputworkspace and it has two input fields where the user can specify the time shift. The user can either specify the time shift in seconds or set an absolute Timestamp to which we want to shift the workspace.
### For Testing:

For checking timestamps in the logs and events there is a [script](http://trac.mantidproject.org/mantid/attachment/ticket/11733/compare_test.py) which should help you.  You can also check this in MantidPlot in the SampleLogs directly and possibly with HDFView.

The recommended workspaces are (They are part of the ExternalData files):
-  SANS2D00022023.nxs 
-  SANS2D00022024.nxs 

It might be worth to use your own MatrixWorkspaces as well.

**Test Dialog Logic**
1. Load SANS2D00022023.nxs into Mantid
2. Open the ChangeTimeZero algorithm 
   - Confirm that by default the RelativeTimeOffset is set to 0.0
   - Confirm that by default the AbsoluteTimeOffset field is empty
3. For the InputWorkspace field select: SANS2D00022023.nxs 
4. For the OutputWorkspace field select: TEST_OUT
5. Set RelativeTimeOffset = 1000.0
6. Set AbsoluteTimeOffset = 2014-10-25T13:58:03
7. Press Run
   - Confirm that the algorithm is aborted

**Test Absolute Time Without Proton Charge Log**
1. Load SANS2D00022024.nxs into Mantid
2. Open the ChangeTimeZero algorithm 
3. For the InputWorkspace field select: SANS2D00022024.nxs 
4. For the OutputWorkspace field select: TEST_OUT
5. Set AbsoluteTimeOffset = 2001-10-25T13:58:03
6. Press Run
   - Confirm that an issue is raised that you cannot use absolute time without having proton_charge in the log files.

**Test Relative Time Offset**
1. Load SANS2D00022023.nxs into Mantid
2. Open the test script
3. For the InputWorkspace field select: SANS2D00022023.nxs 
4. For the OutputWorkspace field select: TEST_OUT
5. Set RelativeTimeOffset = 1000.0
6. Run
   - Confirm that the test scritp does not fail

**Test Absolute Time Offset**
1. Load SANS2D00022023.nxs into Mantid
2. Open the test script
3. For the InputWorkspace field select: SANS2D00022023.nxs 
4. For the OutputWorkspace field select: TEST_OUT
5. Set AbsoluteTimeOffset = 2014-10-25T13:58:03
6. Run
   - Confirm that the test scritp does not fail

**Test Absolute Time Offset With Old Date**
1. Load SANS2D00022023.nxs into Mantid
2. Open the test script
3. For the InputWorkspace field select: SANS2D00022023.nxs 
4. For the OutputWorkspace field select: TEST_OUT
5. Set AbsoluteTimeOffset = 2001-10-25T13:58:03
6. Run
   - Confirm that the test scritp does not fail

**Test Relative Time Offset With Fractional Realtive Time**
1. Load SANS2D00022023.nxs into Mantid
2. Open the test script
3. For the InputWorkspace field select: SANS2D00022023.nxs 
4. For the OutputWorkspace field select: TEST_OUT
5. Set RelativeTimeOffset = 1000.345
6. Press Run
   - Confirm that the test scritp does not fail

**Test Relative Time Offset With Negative Relative Time**
1. Load SANS2D00022023.nxs into Mantid
2. Open the test script
3. For the InputWorkspace field select: SANS2D00022023.nxs 
4. For the OutputWorkspace field select: TEST_OUT
5. Set RelativeTimeOffset =  -1000
6. Press Run
   - Confirm that the test scritp does not fail
